pkg/controllers/teamrbac/teamrolebinding_controller.go

```diff
@@ -481,3 +612,99 @@ func isRoleReferenced(ctx context.Context, c client.Client, teamRoleBinding *gre
 	}
 	return false, nil
 }
+
+// listClusters returns the list of Clusters that match the TeamRoleBinding's ClusterSelector or ClusterName
+func (r *TeamRoleBindingReconciler) listClusters(ctx context.Context, trb *greenhousev1alpha1.TeamRoleBinding) (*greenhousev1alpha1.ClusterList, error) {
+	if trb.Spec.ClusterName != "" {
+		cluster := new(greenhousev1alpha1.Cluster)
+		if err := r.Get(ctx, types.NamespacedName{Name: trb.Spec.ClusterName, Namespace: trb.GetNamespace()}, cluster); err != nil {
+			return nil, err
+		}
+		return &greenhousev1alpha1.ClusterList{Items: []greenhousev1alpha1.Cluster{*cluster}}, nil
+	}
+
+	clusterSelector, err := v1.LabelSelectorAsSelector(&trb.Spec.ClusterSelector)
+	if err != nil {
+		return nil, err
+	}
+	var clusters = new(greenhousev1alpha1.ClusterList)
+	if err := r.List(ctx, clusters, client.InNamespace(trb.GetNamespace()), client.MatchingLabelsSelector{Selector: clusterSelector}); err != nil {
+		return nil, err
+	}
+	return clusters, nil
+}
+
+// initTeamRoleBindingStatus ensures that all required conditions are present in the TeamRoleBinding's Status
+func initTeamRoleBindingStatus(trb *greenhousev1alpha1.TeamRoleBinding) greenhousev1alpha1.TeamRoleBindingStatus {
+	status := trb.Status.DeepCopy()
+	for _, ct := range exposedConditions {
+		if status.GetConditionByType(ct) == nil {
+			status.SetConditions(greenhousev1alpha1.UnknownCondition(ct, "", ""))
+		}
+	}
+	return *status
+}
+
+// setStatus patches the Status of the TeamRoleBinding
+func (r *TeamRoleBindingReconciler) setStatus(ctx context.Context, trb *greenhousev1alpha1.TeamRoleBinding, status greenhousev1alpha1.TeamRoleBindingStatus) error {
+	readyCondition := computeReadyCondition(status)
+	status.StatusConditions.SetConditions(readyCondition)
+	_, err := clientutil.PatchStatus(ctx, r.Client, trb, func() error {
+		trb.Status = status
+		return nil
+	})
+	return err
+}
+
+// computeReadyCondition computes the ReadyCondition based on the TeamRoleBinding's StatusConditions
+func computeReadyCondition(status greenhousev1alpha1.TeamRoleBindingStatus) greenhousev1alpha1.Condition {
+	readyCondition := *status.GetConditionByType(greenhousev1alpha1.ReadyCondition)
+
+	for _, condition := range status.PropagationStatus {
+		if condition.IsTrue() {
+			continue // skip if the RBACReady condition is true for this cluster
+		}
+		readyCondition.Status = metav1.ConditionFalse
+		readyCondition.Message = "Team RBAC reconciliation failed"
+		return readyCondition
+	}
+
+	if status.GetConditionByType(greenhousev1alpha1.ClusterListEmpty).IsTrue() {
```

Comment:
Do we have a case, where we have a not empty `PropagationStatus` with an `ClusterListEmpty == true`?
Then this should maybe be checked first?

Reply:
And putting it to discussion: `ClusterListEmpty==true` could lead to `metav1.CondintionUnkown`, too?
Different to NOT being ready?

Reply:
> And putting it to discussion: ClusterListEmpty==true could lead to metav1.CondintionUnkown, too?
>Different to NOT being ready?

I can't follow, could you elaborate?

Reply:
> Do we have a case, where we have a not empty PropagationStatus with an ClusterListEmpty == true?
> Then this should maybe be checked first?

This would be the case if the `.spec.clusterSelector` is changed and does not match any cluster. In that case the reconciliation will start to clean-up the remote resources. Errors will be reflected in the `.status.propagationStatus` and those lead to a not ready status.

Reply:
> I can't follow, could you elaborate?
Is `readyCondition` really `false` without any clusters to deploy to? I just thought `unkown` might be more appropriate?